src/wallet/wallet.cpp

```diff
@@ -3955,11 +3955,6 @@ std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain& chain, const std::st
                                _("This is the transaction fee you will pay if you send a transaction."));
         }
         walletInstance->m_pay_tx_fee = CFeeRate(nFeePerK, 1000);
-        if (walletInstance->m_pay_tx_fee < chain.relayMinFee()) {
-            error = strprintf(_("Invalid amount for -paytxfee=<amount>: '%s' (must be at least %s)"),
```

Comment:
Addressed in a general comment below

Reply:
In commit "CWallet::Create move chain init message up into calling code" (817308b05e72e875bb968daf31a590eacd993c7d)

I still think this commit is not an improvement and should be dropped because it makes the already complicated handling of these fee options and errors more complicated and less transparent. Instead of handling the settings in one place in `Create`, it splits up handling between two functions `Create` and `AttachChain`, when the code doesn't actually have anything to do with attaching to the chain. It is true that this code **currently** depends on some chain methods. But this should be fixed in the future by [making the chain object less monolithic](https://github.com/bitcoin/bitcoin/blob/741749a6159eb68dd8a3c5ef0d1701337931e6df/src/interfaces/chain.h#L69-L88) and by improving options handling in general.

I understand the motivation for this commit which is stated at https://github.com/bitcoin/bitcoin/pull/20773#issuecomment-760086132 and is basically to help implement the subsequent commit "wallet: make chain optional for CWallet::Create" (37b55f38d31c05f1ae6e22537eea8707463168a6).

But I don't think a good way to make the chain optional is to move these checks to an unrelated place where they don't belong. The simplest and clearest way to make the chain optional for now is simply to skip these checks if the chain pointer is null. And in the future, these checks can be updated to not use the chain interface at all.

Reply:
Hm... thanks for mentioning that the dependency on the chain in accidental. I didn't look deep into it before.

I'm much more happy to drop that commit now, but should we explore other options as well?

It looks like we're just comparing one arg agains other args. paytxfee and maxtxfee against `minRelayFee`, plus `minRelayFee` against some constant. There should be a better place for those checks rather than wallet creation code. I'm not familiar with the codebase enough to say where it belongs.

@ryanofsky and @fjahr could you help me locate a better place for those checks?

Reply:
> @ryanofsky and @fjahr could you help me locate a better place for those checks?

I think the checks are in a good place. Good to validate settings in the same place where they are parsed. There are just other ways this code could access `::minRelayTxFee` value other than by calling a chain method. It could passed as a create option or be part of a different struct or interface. Simply changing the check to `if (chain && walletInstance->m_pay_tx_fee < chain->relayMinFee())` seems good for now though.

Reply:
> And in the future, these checks can be updated to not use the chain interface at all.

For the `chain` topic this seems a good solution going forward.

Otherwise, I am not sure why you are interested in moving this code. These are checks that can make the wallet creation process fail or at least print a warning. So this seems to be a good place for it. Fee options are part of the wallet "world". Maybe the confusion is because these are config options and not args of `createwallet`? Or you don't see the direct connection between fee options and the wallet? These options are only relevant to the wallet and thus are only checked once the user wants to create a wallet because only then weird numbers there can be dangerous for the user. On a smaller, organizational scale some of these checks could probably be helper functions and overall the fee part of wallet create can probably be encapsulated similar to `AttachChain`.

If, however, you want to challenge that fee checks are done at all during wallet creation, that would require a lot more conceptual discussion. After that conceptual discussion, there could be a decision on where to put it. But the chances of a significant change in this regard are slim because this would be a huge behavior change potentially breaking compatibility with any program that uses the wallet because these options will need to be set and checked in some kind of way.

Just some first thoughts by someone also not very familiar with the codebase :) If you have ideas or want to challenge this in general as I teased above I would recommend you to bring that as a topic to one of the next wallet IRC meetings, this Friday for example.

Reply:
> Maybe the confusion is because these are config options and not args of createwallet? 

yes, and 
> and thus are only checked once the user wants to create a wallet.

Shouldn't we check them earlier? They are from the wallet "world" but have nothing to do with the wallet *creation*. Especially `chain.relayMinFee().GetFeePerK() > HIGH_TX_FEE_PER_KB`. It seems just a coincidence that the option parsing is done during wallet creation process.

Nevermind, I've dropped the commit. Thanks for your time and patience.

Reply:
> Shouldn't we check them earlier? They are from the wallet "world" but have nothing to do with the wallet _creation_. 

Yes, these settings could be [`WalletContext`](https://github.com/bitcoin/bitcoin/blob/4da26fb85d6d6d205a2794ccb98beee1302b4a25/src/wallet/context.h#L23) members and shared across multiple `CWallet` instances instead of being `CWallet` members. (Or maybe if we want keep these as global settings but allow individual wallet overrides they could be members of both `CWallet` and `WalletContext`, with `CWallet` copying the `WalletContext` values on construction.) There's a general issue about wallet setting in #13044 for background & discussion about these things. And #19101 is another PR moving more variables to `WalletContext`.

My only reason for wanting to drop the commit is wanting to keep options parsing and option validation code together. I wouldn't want to move the checks without also moving the `IsArgSet`/`GetArg` code they are a part of. But all of this code could run before wallet creation if settings were stored outside `CWallet` in a place like `WalletContext`.

> Especially `chain.relayMinFee().GetFeePerK() > HIGH_TX_FEE_PER_KB`. It seems just a coincidence that the option parsing is done during wallet creation process.

I'm not sure. This is also a warning passed back to the `loadwallet` and `createwallet` RPCs. It seems like a reasonable thing to warn user about when they are loading a wallet, but maybe it could happen at another time. It's more of a UX question than a code organization question, I think.